src/coreComponents/constitutive/fluid/multifluid/compositional/functions/StabilityTest.hpp

```diff
@@ -0,0 +1,231 @@
+      {
+        for( integer const ic : presentComponents )
+        {
+          logTrialComposition[ic] = hyperplane[ic] - logFugacity[ic];
+          trialComposition( trialIndex, ic ) = LvArray::math::exp( logTrialComposition[ic] );
+          normalizedComposition[ic] = trialComposition( trialIndex, ic );
+        }
+        normalizeComposition( numComps, normalizedComposition.toSlice() );
+        EOS_TYPE::computeLogFugacityCoefficients( numComps,
+                                                  pressure,
+                                                  temperature,
+                                                  normalizedComposition.toSliceConst(),
+                                                  componentProperties,
+                                                  logFugacity );
+        real64 error = 0.0;
+        for( integer const ic : presentComponents )
+        {
+          real64 const dG =  logTrialComposition[ic] + logFugacity[ic] - hyperplane[ic];
+          error += (dG*dG);
+        }
+        error = LvArray::math::sqrt( error );
+
+        if( error < MultiFluidConstants::fugacityTolerance )
+        {
+          // Calculate modified tangent plane distance (Michelsen, 1982b) of trial composition relative to input composition
+          double tpd = 1.0;
+          for( integer const ic : presentComponents )
+          {
+            tpd += trialComposition( trialIndex, ic ) * (logTrialComposition[ic] + logFugacity[ic] - hyperplane[ic] - 1.0);
+          }
+          if( tpd < tangentPlaneDistance )
+          {
+            tangentPlaneDistance = tpd;
+          }
+          numberOfStationaryPoints++;
+          break;
+        }
+      }
+    }
+    if( numberOfStationaryPoints == numTrials )
+    {
+      for( integer const ic : presentComponents )
+      {
+        kValues[ic] = trialComposition( 1, ic ) / trialComposition( 0, ic );
+      }
+    }
+    return numberOfStationaryPoints == numTrials;
+  }
+
+private:
+  /**
+   * @brief Calculate which components are present.
+   * @details Creates a list of indices whose components have non-zero mole fraction.
+   * @param[in] numComps number of components
+   * @param[in] composition the composition of the fluid
+   * @param[out] presentComponents the list of present components
+   * @return the number of present components
+   */
+  template< typename ARRAY >
+  GEOS_HOST_DEVICE
+  GEOS_FORCE_INLINE
+  static integer calculatePresentComponents( integer const numComps,
+                                             arraySlice1d< real64 const > const & composition,
+                                             ARRAY & presentComponents )
+  {
+    // Check for machine-zero feed values
+    integer presentCount = 0;
+    for( integer ic = 0; ic < numComps; ++ic )
+    {
+      if( MultiFluidConstants::epsilon < composition[ic] )
+      {
+        presentComponents[presentCount++] = ic;
+      }
+    }
+    presentComponents.resize( presentCount );
```

Comment:
this is not ideal.

Reply:
This is a resize on a `stackArray1d` which I believe shouldn't have much overhead.